### ISSUE TITLE
Possible Regression: NotImplementedException when querying Recursive CTE.

### ISSUE THREAD

**Issue by to11mtm** [linq2db/linq2db#1492]
Describe your issue.

Attempting to Query a CTE I am getting the below exception; This query functioned properly in Linq2DB 2.2.0->2.4.0.

It appears to break in all versions 2.5.1 and up, (i.e. this does seem to work in 2.5.0, but no newer) but I require fixes in newest version for other issues.

```
Exception message:
System.NotImplementedException
The method or operation is not implemented.
Stack trace:
 at LinqToDB.Linq.Builder.SelectContext.ProcessMemberAccess[T](Expression expression, MemberExpression levelExpression, Int32 level, Func`6 action) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\SelectContext.cs:line 983
   at LinqToDB.Linq.Builder.SelectContext.IsExpressionInternal(Expression expression, Int32 level, RequestFor requestFlag) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\SelectContext.cs:line 736
   at LinqToDB.Linq.Builder.SelectContext.IsExpression(Expression expression, Int32 level, RequestFor requestFlag) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\SelectContext.cs:line 632
   at LinqToDB.Linq.Builder.SubQueryContext.IsExpression(Expression expression, Int32 level, RequestFor testFlag) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\SubQueryContext.cs:line 64
   at LinqToDB.Linq.Builder.SubQueryContext.IsExpression(Expression expression, Int32 level, RequestFor testFlag) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\SubQueryContext.cs:line 64
   at LinqToDB.Linq.Builder.ConcatUnionBuilder.UnionContext.BuildExpression(Expression expression, Int32 level, Boolean enforceServerSide) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ConcatUnionBuilder.cs:line 307
   at LinqToDB.Linq.Builder.ExpressionBuilder.TransformExpression(IBuildContext context, Expression expr, Boolean enforceServerSide, String alias) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.QueryBuilder.cs:line 141
   at LinqToDB.Linq.Builder.ExpressionBuilder.<>c__DisplayClass75_0.<BuildExpression>b__0(Expression expr) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.QueryBuilder.cs:line 50
   at LinqToDB.Expressions.Extensions.Transform(Expression expr, Func`2 func) in C:\projects\linq2db\Source\LinqToDB\Expressions\Extensions.cs:line 1354
   at LinqToDB.Expressions.Extensions.<>c__DisplayClass34_0.<Transform>g__Modify|1(MemberBinding b) in C:\projects\linq2db\Source\LinqToDB\Expressions\Extensions.cs:line 1500
   at LinqToDB.Expressions.Extensions.Transform2[T](ICollection`1 source, Func`2 func) in C:\projects\linq2db\Source\LinqToDB\Expressions\Extensions.cs:line 1321
   at LinqToDB.Expressions.Extensions.Transform(Expression expr, Func`2 func) in C:\projects\linq2db\Source\LinqToDB\Expressions\Extensions.cs:line 1546
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildExpression(IBuildContext context, Expression expression, Boolean enforceServerSide) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.QueryBuilder.cs:line 51
   at LinqToDB.Linq.Builder.ConcatUnionBuilder.UnionContext.BuildExpression(Expression expression, Int32 level, Boolean enforceServerSide) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ConcatUnionBuilder.cs:line 294
   at LinqToDB.Linq.Builder.ConcatUnionBuilder.UnionContext.BuildQuery[T](Query`1 query, ParameterExpression queryParameter) in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ConcatUnionBuilder.cs:line 207
   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]() in C:\projects\linq2db\Source\LinqToDB\Linq\Builder\ExpressionBuilder.cs:line 162
   at LinqToDB.Linq.Query`1.CreateQuery(IDataContext dataContext, Expression expr) in C:\projects\linq2db\Source\LinqToDB\Linq\Query.cs:line 279
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr) in C:\projects\linq2db\Source\LinqToDB\Linq\Query.cs:line 232
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache) in C:\projects\linq2db\Source\LinqToDB\Linq\ExpressionQuery.cs:line 84
   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator() in C:\projects\linq2db\Source\LinqToDB\Linq\ExpressionQuery.cs:line 179
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
```

### Steps to reproduce

Apolgies for the size of this repro, I haven't had the opportunity to see if there is a way to minify the issue.

```c#
//Mapping:

public class HierarchyEntryResult
{
	public int RootPortfolioID { get; set; }
	public int PortfolioID { get; set; }
	public int HierarchyLevel { get; set; }
}
public class PrimaryJoinPoint
{
	public int PrimaryId {get;set;}
	public DateTime? ActiveDate { get; set; } //(datetime, null)
	public DateTime? DeactivateDate { get; set; } //(datetime, null)
}
public class SecondaryJoinPoint
{
	public int SecondaryId { get; set; }
	public DateTime? ActiveDate { get; set; } //(datetime, null)
	public DateTime? DeactivateDate { get; set; } //(datetime, null)
}
public class HierarchyDbEntry
{
	public int HierarchyDbEntryId {get;set;}
	public bool IsEntry {get;set;}
}

public class PrimarySecondaryLink
{
	public int LinkID { get; set; } //(int, not null)
	public int? LinkTypeID { get; set; } //(int, null)
	public int ParentID { get; set; } //(int, not null)
	public int ChildID { get; set; } //(int, not null)
	public DateTime? ActiveDate { get; set; } //(datetime, null)
	public DateTime? DeactivateDate { get; set; } //(datetime, null)
}

//Repro Method:
//Call with any Datetime/root ID.
public IEnumerable<HierarchyEntryResult> TestCTEResult(DataConnection conn, DateTime targetDate, int rootId)
{
	return conn.GetCte<HierarchyEntryResult>(theHierarchy =>
					// Anchor:
					(
						from cp in (
					from c in conn.GetTable<PrimaryJoinPoint>()
					where c.ActiveDate <= targetDate
						&& (c.DeactivateDate == null || c.DeactivateDate > targetDate)
					select new HierarchyDbEntry
					{
						HierarchyDbEntryId = c.PrimaryId,
						IsEntry = true
					}
				).Concat(
					from p in conn.GetTable<SecondaryJoinPoint>()
					where p.ActiveDate <= targetDate
						&& (p.DeactivateDate == null || p.DeactivateDate > targetDate)
					select new HierarchyDbEntry
					{
						HierarchyDbEntryId = p.SecondaryId,
						IsEntry = false
					}
				)
						where cp.HierarchyDbEntryId == rootId
						select new HierarchyEntryResult
						{
							RootPortfolioID = rootId,
							PortfolioID = cp.HierarchyDbEntryId,
							HierarchyLevel = 0
						}
					// Recurse:
					).Concat(
						from ph in theHierarchy
						join cpl in conn.GetTable<PrimarySecondaryLink>() on ph.PortfolioID equals cpl.ChildID
						where cpl.ActiveDate <= targetDate
							&& (cpl.DeactivateDate == null || cpl.DeactivateDate > targetDate)
						select new HierarchyEntryResult
						{
							RootPortfolioID = rootId,
							PortfolioID = cpl.ParentID,
							HierarchyLevel = ph.HierarchyLevel + 1
						}
					)
				, "Hierarchy").ToList();
}
```

### Environment details
linq2db version: *2.6.1*
Database Server: *E.g. Sql Server 2012*
Database Provider: *E.g. System.Data.SqlClient 4.0.0*
Operating system: *E.g. Windows 10*
Framework version: *.NET Framework 4.6.2*

**Comment by sdanyliv** [comment]
@to11mtm, thanks for reporting will handle this.

**Comment by MaceWindu** [comment]
Fixed by #1495
